src/source-updater.js

```diff
@@ -24,54 +24,136 @@ const actions = {
   },
   callback: (callback) => (type, sourceUpdater) => {
     callback();
+  },
+  duration: (duration) => (sourceUpdater) => {
+    sourceUpdater.mediaSource.duration = duration;
   }
 };
 
 const updating = (type, sourceUpdater) => {
   const sourceBuffer = sourceUpdater[`${type}Buffer`];
 
-  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queue[type].pending;
+  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queuePending[type];
 };
 
-const shiftQueue = (type, sourceUpdater) => {
-  const queue = sourceUpdater.queue[type];
-
-  if (updating(type, sourceUpdater) || !queue.actions.length || !sourceUpdater.started_) {
-    return;
+const processMediaSourceQueueEntry = (index, sourceUpdater) => {
+  if (updating('audio', sourceUpdater) || updating('video', sourceUpdater)) {
+    // block until all buffers are ready for use
+    return false;
   }
 
-  const action = queue.actions.shift();
+  const queueEntry = sourceUpdater.queue.splice(index, 1)[0];
+
+  queueEntry.action(sourceUpdater);
 
-  queue.pending = action[1];
-  action[0](type, sourceUpdater);
+  if (queueEntry.doneFn) {
+    queueEntry.doneFn();
+  }
+
+  return true;
 };
 
-const pushQueue = (type, sourceUpdater, action) => {
-  const queue = sourceUpdater.queue[type];
+const nextQueueIndexOfType = (type, queue) => {
+  for (let i = 0; i < queue.length; i++) {
+    const queueEntry = queue[i];
 
-  queue.actions.push(action);
-  shiftQueue(type, sourceUpdater);
+    if (queueEntry.type === 'mediaSource') {
+      // If the next entry that uses this type is a media source entry (uses multiple
+      // source buffers), block processing to allow it to go through first.
+      return null;
+    }
+
+    if (queueEntry.type === type) {
+      return i;
+    }
+  }
+
+  return null;
 };
 
-const onUpdateend = (type, sourceUpdater) => () => {
-  const queue = sourceUpdater.queue[type];
+const shiftQueue = (type, sourceUpdater) => {
+  if (sourceUpdater.queue.length === 0) {
+    return;
+  }
+
+  let queueIndex = 0;
+  let queueEntry = sourceUpdater.queue[queueIndex];
 
-  if (!queue.pending) {
-    shiftQueue(type, sourceUpdater);
+  if (queueEntry.type === 'mediaSource') {
+    if (processMediaSourceQueueEntry(queueIndex, sourceUpdater)) {
+      // Only specific source buffer actions must wait for async updateend events. Media
+      // Source actions process synchronously. Therefore, both audio and video source
+      // buffers are now clear to process the next queue entries.
+      shiftQueue('audio', sourceUpdater);
+      shiftQueue('video', sourceUpdater);
+    }
+    return;
   }
 
-  if (!queue.pending) {
-    // nothing in the queue
+  if (type === 'mediaSource') {
```

Comment:
could we just push onto the `queue` in `pushQueue` if it's a mediaSource event instead of calling `shiftQueue`?

Reply:
We have to shift in case nothing is being processed (so that it can process immediately), otherwise it would wait for another action to be pushed onto the queue.